src/frontend/src/optimizer/plan_node/stream_materialize.rs

```diff
@@ -222,7 +222,14 @@ impl StreamMaterialize {
             // No order by for create table, so stream key is identical to table pk.
             (table_pk, pk_column_indices)
         } else {
-            derive_pk(input, user_order_by, &columns)
+            derive_pk(
+                input,
+                user_order_by,
+                &columns,
+                // For index, we can't prune the ORDER KEY,
+                // since it's also the distribution key.
+                table_type != TableType::Index,
```

Comment:
Index actually is also a materialized view. This is a bit special if we can't unify them in this code path. Could you provide an example to illustrate the problem you met? Let's see whether we could resolve it together.

Reply:
This is the failing query.
```
    create table t1 (v1 int, v2 int);
    create materialized view v as select count(*) cnt from t1;
    explain (verbose) create index mv_idx on v(cnt);
```

This is its plan without pruning:
```
    StreamMaterialize { columns: [cnt], stream_key: [], pk_columns: [cnt], pk_conflict: NoCheck }
    └─StreamExchange { dist: HashShard(v.cnt) }
      └─StreamTableScan { table: v, columns: [v.cnt], stream_scan_type: ArrangementBackfill, stream_key: [], pk: [], dist: Single }
```

This is its plan with pruning:
```
    StreamMaterialize { columns: [cnt], stream_key: [], pk_columns: [cnt], pk_conflict: NoCheck }
    └─StreamExchange { dist: HashShard(v.cnt) }
      └─StreamTableScan { table: v, columns: [v.cnt], stream_scan_type: ArrangementBackfill, stream_key: [], pk: [], dist: Single }
```

It has the following functional dependency: `[] -> all_columns`, since it is a singleton.

As mentioned in the PR description, the following section of code:
https://github.com/risingwavelabs/risingwave/blob/a14ff374c624f1a44b38092bc04ef273e03dfd8d/src/frontend/src/handler/create_index.rs#L385-L400

Will result in
1. order key: [cnt]
2. dist key: [0]

Now dist key refers to order key, you can see how it just takes the prefix:
```
        RequiredDist::PhysicalDist(Distribution::HashShard(
            (0..distributed_by_columns_len).collect(),
        )),
```

This seems a little strange, because order key is not appended to pk at this point.
Subsequently, when we derive the pk, that is just stream key combined with order key.
Stream key is `[]`. And after pruning, Order key is `[]` as well. So pk is just `[]`.

Then distribution key is not a subset of `pk` at all.

There's 2 solutions I can think of here:
1. Just avoid index (stupid and simple but not elegant).
2. Separately append distribution key for index elsewhere, instead of relying on order key to be the distribution key.

Reply:
Maybe when we derive a pk, we need to consider the distribution key as well, because theoretically users can specify a distribution key for a materialized view as well, though we don't support it right now, but I remember this feature has been requested several times.

Reply:
BTW, even for the normal materialized view, we need to ensure distribution key is a subset of pk.

Reply:
Fixed it by appending the distribution key if it is not part of the pk.